feat(dspy): added beginner example for multi-input-output

williambrach opened the pull request:
# Added Beginner Multi-Input Multi-Output Example/Tutorial

## Description
Added ew example multi-input and multi-output scenarios in DSPy. 

## Changes Made
- Added new example script (.ipynb) in the `examples/multi-input-output` directory

## Additional Notes
@isaacbmiller , 

Following our discussion on Discord (where my handle is ramang) about adding a beginner-friendly example, I've created this PR. I would greatly appreciate your feedback on both the PR and the example itself. I'm hopeful that this contribution could become part of the official DSPy examples!

isaacbmiller commented on examples/multi-input-output/beginner-multi-input-output.ipynb:
```diff
@@ -0,0 +1,829 @@
+    "        super().__init__()\n",
+    "        # Define the program as a ChainOfThought object\n",
+    "        # self.prog = dspy.PredictiveText(PlagiarismSignature)\n",
+    "        self.prog = dspy.ChainOfThought(PlagiarismSignature)\n",
```
The chain of thought here is actually unnecessary, because you have the explanation field inside of PlagarismSignature

williambrach replied:
I changed the main module in the example to dspy.Predict.

isaacbmiller commented on examples/multi-input-output/beginner-multi-input-output.ipynb:
```diff
@@ -0,0 +1,829 @@
+    "\n",
+    "## TLDR 🚀\n",
+    "\n",
+    "We're going to build a system for code plagiarism detection. Our goal is to compare two input code files, determine if plagiarism has occurred, and provide an explanation for the result. \n",
```
nit: explanation spelled wrong

isaacbmiller commented on examples/multi-input-output/beginner-multi-input-output.ipynb:
```diff
@@ -0,0 +1,829 @@
+    "    def __init__(self) -> None:\n",
+    "        super().__init__()\n",
+    "        # Define the program as a ChainOfThought object\n",
+    "        # self.prog = dspy.PredictiveText(PlagiarismSignature)\n",
```
remove old comment

isaacbmiller commented on examples/multi-input-output/beginner-multi-input-output.ipynb:
```diff
@@ -0,0 +1,829 @@
+    "optimizer = BootstrapFewShot(metric=validate_answer, **config)\n",
+    "optimized_program = optimizer.compile(PlagiarismCoT(), trainset=examples)\n",
+    "score = evaluate(optimized_program)\n",
+    "optimized_program.save(\"BootstrapFewShot_program\")"
```
save it into a json file

williambrach replied:
Updated all save and load programs to use JSON.

isaacbmiller reviewed (changes requested):
Overall looks great! Thank you for putting this together.

My high level thoughts:
- this will be a great addition to DSPy beginner content
- One of the things that people commonly want to see is "what do the optimizers actually do" and "how do they affect the metric"

from this perspective, I would love to see how the score progresses throughout the notebook.

williambrach commented:
> Overall looks great! Thank you for putting this together.
> 
> My high level thoughts:
> 
> * this will be a great addition to DSPy beginner content
> * One of the things that people commonly want to see is "what do the optimizers actually do" and "how do they affect the metric"
> 
> from this perspective, I would love to see how the score progresses throughout the notebook.

I've updated the code to include score printing after each optimizer. I found the extensive optimizer text output less appealing as show in https://github.com/stanfordnlp/dspy/blob/main/examples/qa/hotpot/hotpotqa_with_MIPRO.ipynb. However, if this output is necessary for the process, I can re-run the code and retain the results.

arnavsinghvi11 commented:
This is amazing @williambrach ! really enjoyed the example notebook walkthrough with some great DSPy applications. 

left a couple of proofreading comments on the notebook. 

Also, would it be possible to load in the data externally, and not store it within the repo here? we generally do this with the other datasets in example notebooks and also keeps the repo size manageable. Thanks!

williambrach commented:
> This is amazing @williambrach ! really enjoyed the example notebook walkthrough with some great DSPy applications.
> 
> left a couple of minor proofreading comments in the notebook.
> 
> Also, would it be possible to load the data externally, and not store it within the repo here? we generally do this with the other datasets for example notebooks and also keeps the repo size manageable. Thanks!

Thanks for your response!

I have addressed all the typos. I see that I need to work on proofreading more.

Do you have a preferred way of loading data externally? In the last commit, I updated the loading process to download the dataset from my GitHub repository (e.g., `https://raw.githubusercontent.com/williambrach/LLM-plagiarism-check/main/data/train.tsv`) instead of loading it from disk. Is this okay, or should I upload it to Hugging Face? Can you suggest another approach?

williambrach commented on examples/multi-input-output/beginner-multi-input-output.ipynb:
```diff
@@ -0,0 +1,1093 @@
+    "- Multiple inputs and outputs\n",
+    "- Double validation techniques\n",
+    "\n",
+    "I strongly recommend to read the [DSPy Cheatsheet](https://dspy-docs.vercel.app/docs/cheatsheet) it will help you with quick start.\n",
```
fixed

williambrach commented on examples/multi-input-output/beginner-multi-input-output.ipynb:
```diff
@@ -0,0 +1,1093 @@
+    "   - We are working with a dataset from the publication: [Source Code Plagiarism Detection in Academia with Information Retrieval: Dataset and the Observation](https://github.com/oscarkarnalim/sourcecodeplagiarismdataset/blob/master/IR-Plag-Dataset.zip)\n",
+    "   - We selected a subset and manually labeled the dataset with our output labels. This dataset should be used for training and testing, while the rest of the original dataset should be used for evaluation.\n",
+    "   - Dataset: [train.csv](/data/train.tsv) (65 samples)\n",
+    "   - When you don't hae labeled dataset, it is good idea to try hand-labeling a few examples to get a sense of the task. It will help you to understand the task better and also increse the quality of program.\n",
```
fixed

williambrach commented on examples/multi-input-output/beginner-multi-input-output.ipynb:
```diff
@@ -0,0 +1,1093 @@
+   "source": [
+    "class PlagiarismSignature(dspy.Signature):\n",
+    "    # Clarify something about the nature of the task (expressed below as a docstring)! \n",
+    "    \"\"\"Detect if two code samples are plagiarized. In plagiarized field answer only : Yes if the code samples are plagiarized, No otherwise. In explenation field add the reason why the code samples are/ are not plagiarized.\"\"\"\n",
```
fixed

arnavsinghvi11 commented:
Thanks @williambrach ! Loading through your repo is perfectly fine (will adapt later if needed or feel free to post updates if you move it to HF). Really appreciate this contribution!!
